src/stcal/ramp_fitting/src/slope_fitter.c

```diff
@@ -1173,6 +1227,56 @@ compute_integration_segments(
     return ret;
 }
 
+static int
+cr_list_add(
+        struct cr_list * crs,
+        real_t crmag)
+{
+    struct cr_node * new_node = (struct cr_node*)calloc(1, sizeof(*new_node));
+    const char * msg = "Couldn't allocate memory for cosmic ray node.";
+
+    if (NULL==new_node){
+        PyErr_SetString(PyExc_MemoryError, (const char*)msg);
+        err_ols_print("%s\n", msg);
+        return 1;
+    }
+
+    new_node->crmag = crmag;
+
+    if (0==crs->size) {
+        crs->head = new_node;
+        crs->tail = new_node;
+        crs->size = 1;
+    } else {
+        crs->tail->flink = new_node;
+        crs->tail = new_node;
+        crs->size++;
+    }
```

Comment:
@kmacdonald-stsci I have not used linked lists in c before. I need a quick overview. I googled it and these double linked lists ? I could use a little more detail so I can figure out what it going on. 
Is 'head' the pointer for the VERY FIRST CR segment. And 'tail' the location of the very last one. So are the segments beginning and ends stored or just the first and last locations ? 
In the cr_list_add on line 1248 why is the tail = new_node. I did not think head and tail would equal each other. Is that just for the first cr found. 
Also I am unclear that flink is supposed represent.

Reply:
The linked list I am using for this is a singly linked list.  There are a few different ways to implement this.  My implementation uses two structures, the linked list container and the nodes being listed, with only the forward link being defined in each node.  The forward link is the `flink` parameter that points to the next node in the list.  When `flink` is `NULL`, then then that node is the tail of the list:

https://github.com/spacetelescope/stcal/blob/0f3a23aaa946de203dbf0bbebd1972a29bc87f6e/src/stcal/ramp_fitting/src/slope_fitter.c#L283

The linked list data structure contains useful meta data for the list.  In particular, the head, the tail, and the total number of nodes in each list, which will be used to create the second dimension for the 4-D CRMAG array, which has dimensions `(nints, max_num_crs, nrows, ncols)`.  The tail parameter in the linked list is a convenience variable to go directly to the tail and add the next node, rather than have to traverse the list to the tail each time a new node is added to the list.

https://github.com/spacetelescope/stcal/blob/0f3a23aaa946de203dbf0bbebd1972a29bc87f6e/src/stcal/ramp_fitting/src/slope_fitter.c#L289

So the linked list with _n_ CRs detected looks like this:

```
cr_list->head = cr_node->flink = pointer to the next node
                cr_node->flink = pointer to the next node
                cr_node->flink = pointer to the next node
                cr_node->flink = pointer to the next node
                ...
cr_list->tail = cr_node->flink = NULL
cr->size = _n_
```

So, if some ramp looks like this:

```
data =    [1.,   2.,   13.,      14.,  15.,  16.,  37.,      38.,  39.,  40.,  41.   102.,     103., 104., 105.]
groupdq = [GOOD, GOOD, JUMP_DET, GOOD, GOOD, GOOD, JUMP_DET, GOOD, GOOD, GOOD, GOOD, JUMP_DET, GOOD, GOOD, GOOD]
```

Then the list of CR magnitudes for this ramp will look like this:

```
cr_list->head = {crmag = 11., flink = pointer to next node}
                           {crmag = 21., flink = pointer to next node}
cr_list->tail =    {crmag = 61., flink = NULL}
cr_list->size = 3
```

So, yes, cr_list->head is the very first CR magnitude encountered in the ramp and is the difference between that group data and the previous group data.  The cr_tail->tail is the last known CR magnitude computed, so if another one is found, a new node is allocated and added as the tail.  The result is an ordered linked list starting with the first detected CR magnitude and ending with the last detected CR magnitude.

At line 

https://github.com/spacetelescope/stcal/blob/0f3a23aaa946de203dbf0bbebd1972a29bc87f6e/src/stcal/ramp_fitting/src/slope_fitter.c#L1246

This conditional checks to see if a current list is empty.  If it's empty, the size will be 0.  When a CR is detected, the CR magnitude is computed and a new cr_node is allocated to store that value.  The cr_node is then added to the empty list, resulting in a list of size 1, therefore the head and tail are the same and the size is set to 1.  This list will look like this:

```
cr_list->head = cr_node->flink = NULL
cr_list->tail    = cr_node->flink = NULL
cr_list->size = 1
```

The "else" in this conditional means the list is not empty, so put the newly computed CR magnitude at tail of the list and the size is incremented by 1.